### PR TITLE
fix: stop initializing deployment informer if dynamic sharding is disabled

### PR DESCRIPTION
**Pull request by gdsoumya**
This PR disables deployment informer if dynamic sharding is disabled. This is necessary because the app controller does not need deployment read permissions when running in a rbac constrained setup if user is not using the dynamic sharding feature.

Currently if the controller is missing deployment permission even when dynamic sharding is disabled we can see these error logs from the informer :

```
pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.Deployment: failed to list *v1.Deployment: deployments.apps is forbidden: User "system:serviceaccount:argocd:application-controller" cannot list resource "deployments" in API group "apps" in the namespace "argocd"
```

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by ishitasequeira — COMMENTED** [review]
@gdsoumya , which version of ArgoCD is showing the error? The list permissions were added for the deployment informer in https://github.com/argoproj/argo-cd/pull/16785

**Comment by gdsoumya** [comment]
@ishitasequeira this is in 2.9 and 2.10

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 1]
```diff
@@ -243,11 +245,19 @@ func getClusterSharding(kubeClient *kubernetes.Clientset, settingsMgr *settings.
 	// StatefulSet mode and Deployment mode uses different default values for shard number.
 	defaultShardNumberValue := 0
 	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
```
Comment (ishitasequeira) [inline cmd/argocd-application-controller/commands/argocd_application_controller.go:247]: Can we move this line inside the if block on line 255? This `applicationControllerName` is used only for app controller deployment logic.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
Thanks @gdsoumya for clarifying. Just one minor comment. Overall PR looks good.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go (hunk 1 above)
Comment (gdsoumya) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:247]: Fixed

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 2]
```diff
@@ -238,29 +240,37 @@ func NewCommand() *cobra.Command {
...
+		if err != nil {
+			return nil, fmt.Errorf("(dymanic cluster distribution) failed to get app controller deployment: %v", err)
+		}
+
+		if appControllerDeployment != nil && appControllerDeployment.Spec.Replicas != nil {
+			replicasCount = int(*appControllerDeployment.Spec.Replicas)
+			defaultShardNumberValue = -1
+		} else {
+			return nil, fmt.Errorf("(dymanic cluster distribution) failed to get app controller deployment replica count")
+		}
```
Comment (gdsoumya) [inline cmd/argocd-application-controller/commands/argocd_application_controller.go:264]: This no longer fails silently as the user explicitly opts in to enable the dynamic sharding, any errors in the initialisation step is reported.

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Pull request merged**